lib/seahorn/CexHarness.cc

```diff
@@ -199,6 +303,60 @@ std::unique_ptr<Module> createCexHarness(BmcTraceWrapper &trace,
                          << "Producing harness for " << CF->getName() << "\n";);
           continue;
         }
+        if (CF->getName().equals("memhavoc")) {
+          LOG("cex", errs()
+                         << "Producing harness for " << CF->getName() << "\n";);
+          // previous instruction should be
+          // shadow.mem.load(i32 x, i32 %sm_n, i8* null)
+          if (I.getPrevNonDebugInstruction() == nullptr)
+            continue;
+          const Instruction *prevI = I.getPrevNonDebugInstruction();
+          if (const CallInst *prevCi = dyn_cast<CallInst>(prevI)) {
+            ImmutableCallSite prevCS(prevCi);
+            const Value *prevCV = prevCS.getCalledValue();
+            const Function *prevCF =
+                dyn_cast<Function>(prevCV->stripPointerCasts());
+            if (!(prevCF && prevCF->getName().equals("shadow.mem.load"))) {
+              LOG("cex", errs()
+                             << "Skipping harness for " << CF->getName()
+                             << " because shadow.mem.load cannot be found\n");
+              continue;
+            }
+            // get memhavoc content from second operand of shadow.mem.load
+            const Value *shadowMemPtr = prevCS.getArgOperand(1);
+            const Instruction *shadowMemI = dyn_cast<Instruction>(shadowMemPtr);
+            if (!shadowMemI)
+              continue;
+            Expr shadowMemV = trace.eval(loc, *shadowMemI, true);
+            // get memhavoc size from second operand of memhavoc
+            const Value *size = CS.getArgOperand(1);
+            Expr sizeE;
+            if (auto *sizeI = dyn_cast<Instruction>(size)) {
+              sizeE = trace.eval(loc, *sizeI, true);
+            } else if (auto *sizeConst = dyn_cast<ConstantInt>(size)) {
+              expr::mpz_class sz_mpz = sizeConst->getZExtValue();
+              sizeE = expr::mkTerm<expr::mpz_class>(sz_mpz, trace.efac());
+            } else {
+              LOG("cex", errs() << "unhandled Value of memhavoc size: " << *size
+                                << "\n");
+              continue;
+            }
+            // get info of ptr to havoc
+            const Value *hPtr = CS.getArgOperand(0)->stripPointerCasts();
+            Expr hPtrE;
+            if (auto *hPtrAllocInst = dyn_cast<Instruction>(hPtr)) {
+              hPtrE = trace.eval(loc, *hPtrAllocInst, true);
+            } else {
+              LOG("cex", errs() << "unhandled Value of memhavoc ptr: " << *hPtr
+                                << "\n");
+              continue;
+            }
+
+            FuncValueMap[CF].push_back(shadowMemV);
+            HavocPtrs.push_back(std::make_pair(hPtrE->arg(0), sizeE));
```

Comment:
I think this silently assumes that memory is a struct. This is not always the case.
At any rate, comment what `->arg(0)` is for.

Reply:
has this been addressed or is it obsolete?